src/methods.rs

```diff
@@ -1,18 +1,81 @@
 use rustc_front::hir::*;
 use rustc::lint::*;
 use rustc::middle::ty;
-use rustc::middle::subst::Subst;
+use rustc::middle::subst::{Subst, TypeSpace};
 use std::iter;
 use std::borrow::Cow;
+use std::collections::HashSet;
 
-use utils::{snippet, span_lint, match_path, match_type, walk_ptrs_ty_depth};
+use utils::{snippet, span_lint, match_path, match_type, walk_ptrs_ty_depth,
+    walk_ptrs_ty};
 use utils::{OPTION_PATH, RESULT_PATH, STRING_PATH};
 
 use self::SelfKind::*;
 use self::OutType::*;
 
-#[derive(Copy,Clone)]
-pub struct MethodsPass;
+use rustc::middle::def_id::DefId;
+
+use rustc::middle::ty::TypeFlags;
+
+#[derive(Clone)]
+pub struct MethodsPass { types_implementing_debug: Option<HashSet<DefId>> }
+
+impl MethodsPass {
```

Comment:
Why record all the impls? Instead, just use http://manishearth.github.io/rust-internals-docs/rustc/middle/ty/struct.TraitDef.html#method.for_each_relevant_impl . If the iterator has nonzero length, you're done.

The traitdef already has a hashmap of impls in it, we're really just duplicating that.